pp/wal/hashdex/scraper/encoding.h

```diff
@@ -11,50 +10,27 @@
 namespace PromPP::WAL::hashdex::scraper::encoding {
 enum class SampleValueType : uint8_t { kUint32 = 0b0000'0000, kDouble, kUint8, kUint16, kFloat, kZero, kNaN };
 
-[[nodiscard]] PROMPP_LAMBDA_INLINE inline SampleValueType value_type(const double val) noexcept {
-  if (std::isnan(val)) [[unlikely]] {
-    return SampleValueType::kNaN;
-  }
-  if (val == 0.0) [[unlikely]] {
-    return SampleValueType::kZero;
-  }
-
-  if (std::trunc(val) == val && val > 0.0 && val <= std::numeric_limits<uint32_t>::max()) [[likely]] {
-    const auto uval = static_cast<uint64_t>(val);
-    if (uval <= std::numeric_limits<uint8_t>::max()) {
-      return SampleValueType::kUint8;
-    }
-    if (uval <= std::numeric_limits<uint16_t>::max()) {
-      return SampleValueType::kUint16;
-    }
-    return SampleValueType::kUint32;
-  }
-
-  if (const auto f = static_cast<float>(val); static_cast<double>(f) == val) [[unlikely]] {
-    return SampleValueType::kFloat;
-  }
-
-  return SampleValueType::kDouble;
-}
-
 struct LayoutMarker {
-  uint8_t raw;
+  bool has_ts : 1;
+  uint8_t length_bytes : 3;
+  SampleValueType sample_value_type : 4;
 
-  [[nodiscard]] PROMPP_LAMBDA_INLINE bool has_timestamp() const noexcept { return raw & 0b1000'0000; }
+  [[nodiscard]] PROMPP_LAMBDA_INLINE bool has_timestamp() const noexcept { return has_ts; }
 
-  [[nodiscard]] PROMPP_LAMBDA_INLINE uint8_t count_size_in_bytes() const noexcept { return (raw >> 4) & 0b0000'0111; }
+  [[nodiscard]] PROMPP_LAMBDA_INLINE uint8_t size_length_in_bytes() const noexcept { return length_bytes; }
 
-  [[nodiscard]] PROMPP_LAMBDA_INLINE SampleValueType value_type() const noexcept { return static_cast<SampleValueType>(raw & 0b0000'1111); }
+  [[nodiscard]] PROMPP_LAMBDA_INLINE SampleValueType value_type() const noexcept { return sample_value_type; }
 
   static PROMPP_LAMBDA_INLINE LayoutMarker make(bool has_ts, uint32_t labels_count, SampleValueType value_type) noexcept {
-    const uint8_t bytes_for_count = (std::bit_width(labels_count) + 7) / 8;
-
-    return {static_cast<uint8_t>((has_ts ? 0b1000'0000 : 0) | (bytes_for_count << 4) | (static_cast<uint8_t>(value_type) & 0b0000'1111))};
+    const uint8_t bytes_for_count = BareBones::Bit::to_ceil_bytes(std::bit_width(labels_count));
+    return LayoutMarker{.has_ts = has_ts, .length_bytes = bytes_for_count, .sample_value_type = value_type};
   }
 };
 
 class SampleCodec {
  public:
+  static constexpr size_t kPreallocatedWriteSize = sizeof(Primitives::Sample);
```

Comment:
```suggestion
  static constexpr size_t kMaximumEncodingSize = sizeof(Primitives::Sample);
```
Because codec doesn't need to know about preallocation